### PR TITLE
Don't add rpaths to system libraries

### LINKED ISSUE #2661: Assertion error when using libm from find_library()
Running Meson fails with Nautilus, as we do explicit linking with libm.

```
Traceback (most recent call last):
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/mesonmain.py", line 353, in run
    app.generate()
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/mesonmain.py", line 152, in generate
    self._generate(env)
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/mesonmain.py", line 209, in _generate
    g.generate(intr)
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/backend/ninjabackend.py", line 210, in generate
    self.generate_target(t, outfile)
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/backend/ninjabackend.py", line 461, in generate_target
    elem = self.generate_link(target, outfile, outname, obj_list, linker, pch_objects)
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/backend/ninjabackend.py", line 2465, in generate_link
    self.determine_rpath_dirs(target),
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/backend/backends.py", line 337, in determine_rpath_dirs
    for rp in self.rpaths_for_bundled_shared_libraries(target):
  File "/home/ernestas/Development/github.com/mesonbuild/meson/mesonbuild/backend/backends.py", line 326, in rpaths_for_bundled_shared_libraries
    assert(not os.path.isabs(rel_to_src))
AssertionError
```

The code in Meson checks if the path is a system path by looking at its prefix, but on my machine it seems to be a bit more exotic: `/usr/bin/../lib/gcc/x86_64-redhat-linux/7/../../../../lib64/libm.so`.

### PR DESCRIPTION
**Pull request by NickeZ**
Fixes #2661

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -317,8 +317,8 @@ def rpaths_for_bundled_shared_libraries(self, target):
                 if len(la) == 1 and os.path.isabs(la[0]):
                     # The only link argument is an absolute path to a library file.
                     libpath = la[0]
-                    if libpath.startswith(('/usr/lib', '/lib')):
-                        # No point in adding system paths.
+                    if(!absdir.startswith(self.environment.get_source_dir())):
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

[E999] SyntaxError: invalid syntax

**Comment by jpakkane** [comment]
We need to add paths to libraries in `/usr/local/lib` because otherwise they stop working.

**Comment by nirbheek** [comment]
I think a better way to fix #2661 is to use `pathlib.PurePath()` which will simplify that weird complicated path and yield `/usr/lib/`.

**Comment by NickeZ** [comment]
I'm not so sure PurePath helps in that case. Anyway the real problem is that the code assumes that one string is a substring of another string. Which it isn't always. I made a new try, I'll rebase and squash if you approve it.

```
>>> PurePath('foo/../bar')
PurePosixPath('foo/../bar')

(a naïve approach would make PurePosixPath('foo/../bar') equivalent to PurePosixPath('bar'), which is wrong if foo is a symbolic link to another directory)
```

**Comment by textshell** [comment]
I was just stumbling about a related problem, where i want to use a library that is not from the source dir and not in a system path. This patch fixes the path related part of my problem.

As for filtering of system paths. Would not `os.path.realpath` be an option to resolve complex relative paths the right way? That way all system pathes would need to be `realpath`·ed too. But that should work.

**Comment by NickeZ** [comment]
Should we call `os.path.realpath` on `la[0]`?

**Comment by textshell** [comment]
I was thinking something like this
```
    if os.path.dirname(os.path.realpath(la[0])) not in (os.path.realpath('/lib'), os.path.realpath('/usr/lib'), ...):
```
Not sure what the list of blacklisted pathes should be. I don't think `/usr/lib/foobar/foobar.so` should be covered by the blacklist, when `/usr/lib/foobar/` is not in the system search path. But multiarch pathes likely should be blacklisted too...

**Comment by NickeZ** [comment]
I guess it should be a system config to specify which paths to exclude. Maybe it is even possible to ask `ld` about the paths

**Comment by NickeZ** [comment]
https://stackoverflow.com/questions/9922949/how-to-print-the-ldlinker-search-path

Should we ask ldconfig or gcc/ld?

**Comment by textshell** [comment]
I'm not sure it's worth the time to get these from external programs. That would need to be cached, etc.
Maybe start with a hardcoded list and see if @jpakkane wants the added complexity of actually getting the systems search list?

**Comment by NickeZ** [comment]
Below is what I get on my machine. Should I add them all or do we stick to the current two (`/lib`, `/usr/lib`)?

```
gcc -m64 -Xlinker --verbose 2>/dev/null
SEARCH_DIR("=/lib/x86_64-linux-gnu");
SEARCH_DIR("=/usr/lib/x86_64-linux-gnu");
SEARCH_DIR("=/lib64");
SEARCH_DIR("=/usr/lib64");
SEARCH_DIR("=/lib");
SEARCH_DIR("=/usr/lib");
SEARCH_DIR("=/usr/x86_64-linux-gnu/lib64");
SEARCH_DIR("=/usr/x86_64-linux-gnu/lib");

gcc -m32 -Xlinker --verbose 2>/dev/null
SEARCH_DIR("=/lib/i386-linux-gnu");
SEARCH_DIR("=/usr/lib/i386-linux-gnu");
SEARCH_DIR("=/lib32");
SEARCH_DIR("=/usr/lib32");
SEARCH_DIR("=/lib");
SEARCH_DIR("=/usr/lib");
SEARCH_DIR("=/usr/i386-linux-gnu/lib32");
SEARCH_DIR("=/usr/x86_64-linux-gnu/lib32");
SEARCH_DIR("=/usr/i386-linux-gnu/lib");
```

**Comment by nirbheek** [comment]
> Maybe it is even possible to ask ld about the paths

See: `self.get_library_dirs()` (compilers/c.py). The problem is that the search paths used by the runtime linker may not be the same as `ld`. But using `ld` paths is definitely better than hard-coding them.

**Comment by jpakkane** [comment]
If we ignore evrything starting with `/lib` or `/usr/lib` does anything break? If we can make this work with a static test, we should because simpler is better.

**Comment by nirbheek** [comment]
This would not be covered: 

```
SEARCH_DIR("=/usr/x86_64-linux-gnu/lib64");
```

I think being correct here is important. If it leads to slowdowns, we can cache it on the compiler object.

**Comment by jpakkane** [comment]
Is that an actual thing that some distro is actually using? I've never seen that in the wild.

**Comment by textshell** [comment]
I was suggesting using `os.path.realpath` because of concerns regarding symlinks in complicated path strings. One downside of using realpath is that you end up with pathes that might look different than expected so subdirectory relationships are hard to cleanly detect. (i.e. realpath('/lib/') == '/usr/lib' on systems with usr-merge, i.e. fedora/arch). Thus the most correct way is to exact match with a list of also realpath·ed locations. Of course meson could simply say that having symlinks to directories in /usr/lib and /lib is not supported in all cases (i.e. meson might generate rpath if '/usr/lib/x86_64-linux-gnu' is a symlink to something outside of '/usr/lib').

Other points
* '/usr/lib/somepackage' usually is not in the dynamic linker searchlist, so it can be argued that just looking for a prefix match is wrong.
* When compiling 32bit code on "Legacy biarch" '/usr/lib32' (or '/usr/libx32') might be the pathes that should be excluded.

But i think it's not really needed to make a perfect fix in this PR. Also i really also prefer a simple solution, with might be either to include a decent list of "what's likely to be in the system dynamic linker path if it exists" or go with a simple but non perfect prefix matching (possibly after realpath or `os.path.normpath` depending on what kind of symlinks tricks we care to support). Esp. as all get real dynamic linker settings approaches are likely highly platform specific.

**Comment by NickeZ** [comment]
I squashed the commits and ignored the blacklist for now. This fixes the most pressing issue for me which is that meson wasn't working on windows.

**Comment by jpakkane** [comment]
The current solution seems to accept everything outside src dir. If we include things starting with `/usr/lib` and `/lib` it should not break anything (but may leave some extra rpaths hanging around). So it should be safe but not perfect?

**Comment by textshell** [comment]
@jpakkane In some other issue you said that some distros are really picky about rpath entries. So i think it would be good to have a blacklist that covers all default directories where those distros ship libraries.

Esp, since i plan proposing (in a future PR) to allow this logic to also work with pkg-config found libraries. But i'm happy to solve that in my pull request, keeping this one as simple as possible.

**Comment by jpakkane** [comment]
> some distros are really picky about rpath entries

These rpaths are only used inside the build dir. When installing they are stripped away. The only rpath entry that remains is the one that the user specifies with `install_rpath`.

**Comment by textshell** [comment]
Oh, this is only used for the build directory rpath, right. So it really doesn't matter much if this adds too many rpath. So that only leaves not adding an rpath for libraries in '/usr/lib/something' which seems to be better solved in a different PR if at all.

**Comment by albertvaka** [comment]
Any ETA for the 0.44.1 release?

**Comment by jpakkane** [comment]
@nirbheek?

**Comment by nirbheek** [comment]
> Any ETA for the 0.44.1 release?

Things are in good shape for the branch, we can release early next week: Tue/Wed.

**Pull request merged**